Clear global environment and load libraries
```{r}
rm(list=ls())
library(CellChat)
library(patchwork)
library(dplyr)
library(Seurat)
options(stringsAsFactors = FALSE)
```

Read in Sequencing Matrix
```{r}
file_name <- "LCMA01_BEVR_CD31"
position <- 1
seurats <- c()

make_Seurat <- function(file_name){
  directory_string <- paste("C:/Users/19193/OneDrive - The University of Chicago",
                            "/Documents/UChicago/Internships etc/",
                            "Internship Opportunities/Aghi Lab/CD31 sorted/",
                            file_name, "/filtered_feature_bc_matrix/",
                            sep = "")
  
  
  assign(paste0(file_name, "_Mtx"), ReadMtx(
    mtx = paste(directory_string, "matrix.mtx.gz", sep = ""),
    features = paste(directory_string, "features.tsv.gz", sep = ""),
    cells = paste(directory_string, "barcodes.tsv.gz", sep = ""))) 
  
  append(seurats, assign(file_name, CreateSeuratObject(counts = LCMA01_BEVR_CD31_Mtx)))
}

make_Seurat(file_name)

```

Seurat Data Prep
```{r}
for()

LCMA01_BEVR_CD31 <- subset(LCMA01_BEVR_CD31, subset= nFeature_RNA > 200 & nFeature_RNA < 2500)

LCMA01_BEVR_CD31 <- NormalizeData(LCMA01_BEVR_CD31, normalization.method = "LogNormalize", scale.factor = 10000)

LCMA01_BEVR_CD31 <- FindVariableFeatures(LCMA01_BEVR_CD31, selection.method = "vst", nfeatures=2000)

all.genes <- rownames(LCMA01_BEVR_CD31)

LCMA01_BEVR_CD31 <- ScaleData(LCMA01_BEVR_CD31, features = all.genes, block.size = 1000)
```
Create Clusters
```{r}
LCMA01_BEVR_CD31 <- RunPCA(LCMA01_BEVR_CD31,  features=VariableFeatures(object=LCMA01_BEVR_CD31))
print(LCMA01_BEVR_CD31[["pca"]], dims=1:15, nfeatures = 15)

LCMA01_BEVR_CD31 <- FindNeighbors(LCMA01_BEVR_CD31, dims=1:10)
LCMA01_BEVR_CD31 <- FindClusters(LCMA01_BEVR_CD31, resolution=0.5)

LCMA01_BEVR_CD31 <- RunUMAP(LCMA01_BEVR_CD31, dims=1:10)
DimPlot(LCMA01_BEVR_CD31, reduction="umap")
```

Create Cell Chat
```{r}
#Identify Cluster Markers

# for(i in 1:13){
#   markerName <- paste("marker", intToUtf8(i+64), sep="")
#   assign(markerName, FindMarkers(LCMA01_BEVR_CD31, ident.1=i-1, min.pct=0.1))
# }
# markerA <- FindMarkers(LCMA01_BEVR_CD31, ident.1=0, min.pct=0.1)
# markerB <- FindMarkers(LCMA01_BEVR_CD31, ident.1=1, min.pct=0.1)
# markerC <- FindMarkers(LCMA01_BEVR_CD31, ident.1=2, min.pct=0.1)
# markerD <- FindMarkers(LCMA01_BEVR_CD31, ident.1=3, min.pct=0.1)
# markerE <- FindMarkers(LCMA01_BEVR_CD31, ident.1=4, min.pct=0.1)
# markerF <- FindMarkers(LCMA01_BEVR_CD31, ident.1=5, min.pct=0.1)
# markerG <- FindMarkers(LCMA01_BEVR_CD31, ident.1=6, min.pct=0.1)
# markerH <- FindMarkers(LCMA01_BEVR_CD31, ident.1=7, min.pct=0.1)
# markerI <- FindMarkers(LCMA01_BEVR_CD31, ident.1=8, min.pct=0.1)
# markerJ <- FindMarkers(LCMA01_BEVR_CD31, ident.1=9, min.pct=0.1)
# markerK <- FindMarkers(LCMA01_BEVR_CD31, ident.1=10, min.pct=0.1)
# markerL <- FindMarkers(LCMA01_BEVR_CD31, ident.1=11, min.pct=0.1)
# markerM <- FindMarkers(LCMA01_BEVR_CD31, ident.1=12, min.pct=0.1)



#Assign New Cluster Names

LCMA01_BEVR_CD31$seurat_clusters <- as.factor(as.numeric(as.character(LCMA01_BEVR_CD31$seurat_clusters)) + 1)

#Make Cell Chat Object
LCMA01_BEVR_CD31_Chat <- createCellChat(object = LCMA01_BEVR_CD31, group.by="seurat_clusters")
```

```{r}


"Set ligand-receptor interaction db"
CellChatDB <- CellChatDB.human
showDatabaseCategory(CellChatDB)
dplyr::glimpse(CellChatDB$interaction)

# use a subset of CellChatDB for cell-cell communication analysis
# CellChatDB.use <- subsetDB(CellChatDB, search = "Secreted Signaling") # use Secreted Signaling
# use all CellChatDB for cell-cell communication analysis
CellChatDB.use <- CellChatDB # simply use the default CellChatDB

# set the used database in the object
LCMA01_BEVR_CD31_Chat@DB <- CellChatDB.use
```

```{r}
# subset the expression data of signaling genes for saving computation cost
LCMA01_BEVR_CD31_Chat <- subsetData(LCMA01_BEVR_CD31_Chat) # This step is necessary even if using the whole database
LCMA01_BEVR_CD31_Chat <- identifyOverExpressedGenes(LCMA01_BEVR_CD31_Chat)
LCMA01_BEVR_CD31_Chat <- identifyOverExpressedInteractions(LCMA01_BEVR_CD31_Chat)

# project gene expression data onto PPI (Optional: when running it, USER should set `raw.use = FALSE` in the function `computeCommunProb()` in order to use the projected data)
# LCMA01_BEVR_CD31_Chat <- projectData(LCMA01_BEVR_CD31_Chat, PPI.human)

```

```{r}
future::plan("multiprocess", workers = 4)
LCMA01_BEVR_CD31_Chat <- computeCommunProb(LCMA01_BEVR_CD31_Chat)
# Filter out the cell-cell communication if there are only few number of cells in certain cell groups
LCMA01_BEVR_CD31_Chat <- filterCommunication(LCMA01_BEVR_CD31_Chat, min.cells = 10)

```

```{r}
LCMA01_BEVR_CD31_Chat_DF_LR <- subsetCommunication(LCMA01_BEVR_CD31_Chat)
LCMA01_BEVR_CD31_Chat_DF_Net <- subsetCommunication(LCMA01_BEVR_CD31_Chat, slot.name = "netP")
```

```{r}
LCMA01_BEVR_CD31_Chat <- computeCommunProbPathway(LCMA01_BEVR_CD31_Chat)
```

```{r}
#Generate interaction network
LCMA01_BEVR_CD31_Chat <- aggregateNet(LCMA01_BEVR_CD31_Chat)

#Visualize interaction network
groupSize <- as.numeric(table(LCMA01_BEVR_CD31_Chat@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(LCMA01_BEVR_CD31_Chat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(LCMA01_BEVR_CD31_Chat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
```

```{r}
#Visualize MHC-II pathway interactions
netVisual_circle(LCMA01_BEVR_CD31_Chat@netP$prob[,,"MHC-II"])
```

```{r}
#Generate Circle Plots
LCMA01_BEVR_CD31_Chat <- netAnalysis_computeCentrality(LCMA01_BEVR_CD31_Chat, slot.name = "netP")

par(mfrow=c(5,5))
par(mai=c(0.1, 0.1, 0.1, 0.1))
for (pathway in LCMA01_BEVR_CD31_Chat@netP$pathways){
  # print(pathway)
  # netVisual_circle(LCMA01_BEVR_CD31_Chat@netP$prob[,,pathway], title.name = pathway)
  netAnalysis_signalingRole_network(LCMA01_BEVR_CD31_Chat, signaling=pathway)
}
```

```{r}

LCMA01_BEVR_CD31_Chat <- computeNetSimilarity(LCMA01_BEVR_CD31_Chat, type = "functional")
LCMA01_BEVR_CD31_Chat <- netEmbedding(LCMA01_BEVR_CD31_Chat, type = "functional")
LCMA01_BEVR_CD31_Chat <- netClustering(LCMA01_BEVR_CD31_Chat, type = "functional")
# Visualization in 2D-space
netVisual_embedding(LCMA01_BEVR_CD31_Chat, type = "functional", label.size = 3.5)
# netVisual_embeddingZoomIn(LCMA01_BEVR_CD31_Chat, type = "functional", nCol = 2)
```

```{r}
## Visualize cell-cell communication mediated by multiple ligand-receptors or signaling pathways
# show all the significant interactions (L-R pairs) from some cell groups (defined by 'sources.use') to other cell groups (defined by 'targets.use')
netVisual_bubble(LCMA01_BEVR_CD31_Chat, sources.use = c(1:12), targets.use = c(1:12), remove.isolate = FALSE)
```

```{r}
netVisual_chord_gene(LCMA01_BEVR_CD31_Chat, sources.use = 3, targets.use = 3, signaling = LCMA01_BEVR_CD31_Chat@netP$pathways)
```



